src/vs/workbench/api/node/extHostLanguageFeatures.ts

```diff
@@ -285,11 +285,37 @@ class QuickFixAdapter {
 			}
 		});
 
-		return asWinJsPromise(token => this._provider.provideCodeActions(doc, ran, { diagnostics: allDiagnostics }, token)).then(commands => {
+		return asWinJsPromise(token =>
+			this._provider.provideCodeActions2
+				? this._provider.provideCodeActions2(doc, ran, { diagnostics: allDiagnostics }, token)
+				: this._provider.provideCodeActions(doc, ran, { diagnostics: allDiagnostics }, token)
+		).then(commands => {
 			if (!Array.isArray(commands)) {
 				return undefined;
 			}
-			return commands.map(command => this._commands.toInternal(command));
+			return commands.map(action => {
+				if (!action) {
+					return undefined;
+				}
+
+				if (typeof action.command === 'string') {
+					return this._commands.toInternal(action as vscode.Command);
+				}
+
+				const codeAction = action as vscode.CodeAction;
+				return {
+					title: codeAction.title,
+					command: codeAction.command ? this._commands.toInternal(codeAction.command) : undefined,
+					edits: codeAction.edits
```

Comment:
For the sake of simplicity we can also make `edits` always a `WorkspaceEdit`. We'd treat the `TextEdit` as sugar and translate it into a workspace edit... Makes the internal CodeAction simpler and easier to handle.